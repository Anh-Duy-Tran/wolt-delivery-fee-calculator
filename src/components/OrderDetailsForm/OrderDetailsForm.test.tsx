import { expect, it, describe, vi, beforeEach } from 'vitest';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { OrderDetailsForm } from '.';

type FormFields =
  | 'cartValue'
  | 'deliveryDistance'
  | 'numberOfItems'
  | 'orderTime';

const VALID_FORM_DATA: Record<FormFields, string> = {
  cartValue: '100',
  deliveryDistance: '500',
  numberOfItems: '3',
  orderTime: '2024-01-22T12:00',
};

const INVALID_FORM_DATA: Record<FormFields, string> = {
  cartValue: '-100',
  deliveryDistance: '500.5',
  numberOfItems: '0',
  orderTime: '2024-02-31T12:00',
};

describe('OrderDetailsForm Component', () => {
  const mockHandleResult = vi.fn();
  const mockSaveCalculatedOrderDetails = vi.fn();

  beforeEach(() => {
    render(
      <OrderDetailsForm
        handleResult={mockHandleResult}
        saveCalculatedOrderDetails={mockSaveCalculatedOrderDetails}
      />
    );
  });

  it('should render correct to the snapshot', () => {
    const { asFragment } = render(
      <OrderDetailsForm
        handleResult={mockHandleResult}
        saveCalculatedOrderDetails={mockSaveCalculatedOrderDetails}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders all input fields', () => {
    expect(screen.getByTestId('cartValue')).toBeInTheDocument();
    expect(screen.getByTestId('deliveryDistance')).toBeInTheDocument();
    expect(screen.getByTestId('numberOfItems')).toBeInTheDocument();
    expect(screen.getByTestId('orderTime')).toBeInTheDocument();
  });

  it('displays validation errors for empty fields', async () => {
    fireEvent.click(screen.getByText(/calculate fee/i));
    await waitFor(() => {
      expect(screen.getByText(/cart value is required/i)).toBeInTheDocument();
      expect(
        screen.getByText(/delivery distance is required/i)
      ).toBeInTheDocument();
      expect(
        screen.getByText(/number of items is required/i)
      ).toBeInTheDocument();
      expect(
        screen.getByText(/please enter a valid time/i)
      ).toBeInTheDocument();
    });
  });

  it('should not submit form with invalid data', async () => {
    // Try to set one of the field in the form to the invalid data and tried to submit the form
    Object.entries(INVALID_FORM_DATA).forEach(
      ([invalidDataField, invalidData]) => {
        // Set the form field to the valid data except for one of the field with invalid data
        Object.entries(VALID_FORM_DATA).forEach(([dataField, data]) =>
          fireEvent.change(screen.getByTestId(dataField), {
            target: {
              value: dataField === invalidDataField ? invalidData : data,
            },
          })
        );

        fireEvent.click(screen.getByText(/calculate fee/i));
      }
    );

    await waitFor(() => {
      expect(mockSaveCalculatedOrderDetails).not.toHaveBeenCalled();
      expect(mockHandleResult).not.toHaveBeenCalled();
    });
  });

  it('should submits form with valid data', async () => {
    fireEvent.change(screen.getByTestId('cartValue'), {
      target: { value: VALID_FORM_DATA.cartValue },
    });
    fireEvent.change(screen.getByTestId('deliveryDistance'), {
      target: { value: VALID_FORM_DATA.deliveryDistance },
    });
    fireEvent.change(screen.getByTestId('numberOfItems'), {
      target: { value: VALID_FORM_DATA.numberOfItems },
    });
    fireEvent.change(screen.getByTestId('orderTime'), {
      target: { value: VALID_FORM_DATA.orderTime },
    });

    fireEvent.click(screen.getByText(/calculate fee/i));

    await waitFor(() => {
      expect(mockSaveCalculatedOrderDetails).toHaveBeenCalledWith({
        cartValue: parseFloat(VALID_FORM_DATA.cartValue),
        deliveryDistance: parseInt(VALID_FORM_DATA.deliveryDistance, 10),
        numberOfItems: parseInt(VALID_FORM_DATA.numberOfItems, 10),
        orderTime: new Date(VALID_FORM_DATA.orderTime),
      });
      expect(mockHandleResult).toHaveBeenCalledTimes(1);
    });
  });
});
