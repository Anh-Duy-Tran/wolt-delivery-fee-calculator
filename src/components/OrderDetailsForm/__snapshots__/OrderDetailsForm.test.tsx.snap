// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OrderDetailsForm Component > should render correct to the snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col gap-4 w-full rounded-xl bg-white p-5 shadow-xl"
    data-testid="card"
  >
    <h2>
      ORDER INFORMATION
    </h2>
    <form
      class="flex flex-col gap-2"
      novalidate=""
    >
      <div>
        <div
          class="relative flex w-full items-center justify-center"
        >
          <input
            autocomplete="off"
            class="peer w-full rounded-md border-2 border-wolt-grey p-4 pb-[6px] pt-[22px] outline-none transition placeholder:opacity-0 hover:border-wolt-blue focus:border-wolt-blue focus:shadow-border-like focus:placeholder:opacity-100"
            data-testid="cartValue"
            inputmode="numeric"
            name="cartValue"
            step="0.1"
            type="text"
          />
          <label
            class="linear duration-config pointer-events-none absolute w-full origin-left translate-x-4 scale-100 
        select-none opacity-60 transition-all peer-focus:-translate-y-3 peer-focus:scale-[65%] false"
          >
            Cart Value (EUR)
          </label>
        </div>
      </div>
      <div>
        <div
          class="relative flex w-full items-center justify-center"
        >
          <input
            autocomplete="off"
            class="peer w-full rounded-md border-2 border-wolt-grey p-4 pb-[6px] pt-[22px] outline-none transition placeholder:opacity-0 hover:border-wolt-blue focus:border-wolt-blue focus:shadow-border-like focus:placeholder:opacity-100"
            data-testid="deliveryDistance"
            inputmode="numeric"
            name="deliveryDistance"
            step="10"
            type="text"
          />
          <label
            class="linear duration-config pointer-events-none absolute w-full origin-left translate-x-4 scale-100 
        select-none opacity-60 transition-all peer-focus:-translate-y-3 peer-focus:scale-[65%] false"
          >
            Delivery Distance (Meter)
          </label>
        </div>
      </div>
      <div>
        <div
          class="relative flex w-full items-center justify-center"
        >
          <input
            autocomplete="off"
            class="peer w-full rounded-md border-2 border-wolt-grey p-4 pb-[6px] pt-[22px] outline-none transition placeholder:opacity-0 hover:border-wolt-blue focus:border-wolt-blue focus:shadow-border-like focus:placeholder:opacity-100"
            data-testid="numberOfItems"
            inputmode="numeric"
            name="numberOfItems"
            type="text"
          />
          <label
            class="linear duration-config pointer-events-none absolute w-full origin-left translate-x-4 scale-100 
        select-none opacity-60 transition-all peer-focus:-translate-y-3 peer-focus:scale-[65%] false"
          >
            Number of Items
          </label>
        </div>
      </div>
      <div>
        <div
          class="relative flex w-full items-center justify-center"
        >
          <input
            class="peer w-full rounded-md border-2 border-wolt-grey p-4 pb-[6px] pt-[22px] outline-none transition placeholder:opacity-0 hover:border-wolt-blue focus:border-wolt-blue focus:shadow-border-like focus:placeholder:opacity-100"
            data-testid="orderTime"
            name="orderTime"
            placeholder="DD.MM.YYYY"
            type="datetime-local"
          />
          <label
            class="linear duration-config pointer-events-none absolute w-full origin-left translate-x-4 scale-100 
        select-none opacity-60 transition-all peer-focus:-translate-y-3 peer-focus:scale-[65%] -translate-y-3 scale-[65%]"
          >
            Time
          </label>
        </div>
      </div>
      <button
        class="mt-2 rounded-[6px] bg-wolt-blue p-3 text-white"
        data-testid="calculateFee"
        type="submit"
      >
        CALCULATE FEE
      </button>
    </form>
  </div>
</DocumentFragment>
`;
